### DISCUSSION TITLE
Regex pattern in autodiscoverFilter

### DISCUSSION THREAD

**Discussion by tonnico** [renovatebot/renovate#20548] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

34.134.1

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

I would like to filter a specific subgroup in gitlab. Is it possible somehow with the `autodiscoverFilter` feature?

Something like?
```json
"autodiscoverFilter": ["/group/(?!dontincludeme).*/"]
```

### What you tried so far.

see above, but it failed to parse.

```
Fatal error: Failed to parse regex pattern "/group/(?!dontincludeme).*/"
```

### Relevant debug logs

<details><summary>Logs</summary>

```
       "loggerErrors": [
         {
           "name": "renovate",
           "level": 60,
           "logContext": "QCqV9uuDI5oOaHek99DhP",
           "err": {
             "message": "Failed to parse regex pattern \"/group/(?!dontincludeme).*/\"",
             "stack": "Error: Failed to parse regex pattern \"/group/(?!dontincludeme).*/\"\n    at applyFilters (/opt/buildpack/tools/renovate/34.134.1/node_modules/renovate/lib/workers/global/autodiscover.ts:89:15)\n    at autodiscoverRepositories (/opt/buildpack/tools/renovate/34.134.1/node_modules/renovate/lib/workers/global/autodiscover.ts:35:18)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at start (/opt/buildpack/tools/renovate/34.134.1/node_modules/renovate/lib/workers/global/index.ts:148:14)\n    at /opt/buildpack/tools/renovate/34.134.1/node_modules/renovate/lib/renovate.ts:18:22"
           },
           "msg": "Fatal error: Failed to parse regex pattern \"/group/(?!dontincludeme).*/\""
         }
       ]
```

</details>

**Comment by RahulGautamSingh** [comment, marked as answer]
Renovate uses RE2 which doesn't support negative lookaheads.

**Reply by tonnico** [reply]
Didn't see that it's RE2 (or overlooked). Thanks.
My solution for a simple exclude `dont`:
~`"autodiscoverFilter": ["/group/[^(dontincludeme)].*/"]`~
`"autodiscoverFilter": ["/group/([^d]|d[^o]|do[^n]|don[^t]).*/"]`

**Reply by szpak** [reply]
Thanks @tonnico! It helped me solve problem with "manual sharding" (better than GPT ;-) ).

**Reply by szpak** [reply]
Btw (for those hitting this discussion in the future), with Renovate 40+, it should be possible to do it in the easier way:
https://github.com/renovatebot/renovate/releases/tag/40.0.0
https://github.com/renovatebot/renovate/issues/21347
